src/history/hs_rec.c

```diff
@@ -605,25 +614,33 @@ __wt_hs_insert_updates(
              * Calculate reverse modify and clear the history store records with timestamps when
              * inserting the first update. Always write on-disk data store updates to the history
              * store as a full update because the on-disk update will be the base update for all the
-             * updates that are older than the on-disk update.
+             * updates that are older than the on-disk update. Limit the number of consecutive
+             * reverse modifies for standard updates. We want to ensure we do not store a large
+             * chain of reverse modifies as to impact read performance.
              *
              * Due to concurrent operation of checkpoint and eviction, it is possible that history
              * store may have more recent versions of a key than the on-disk version. Without a
              * proper base value in the history store, it can lead to wrong value being restored by
              * the RTS.
              */
             nentries = MAX_REVERSE_MODIFY_NUM;
-            if (!F_ISSET(upd, WT_UPDATE_DS) && upd->type == WT_UPDATE_MODIFY &&
+            if (!F_ISSET(upd, WT_UPDATE_DS) && !F_ISSET(prev_upd, WT_UPDATE_DS) &&
               enable_reverse_modify &&
+              (upd->type == WT_UPDATE_MODIFY ||
+                total_nb_modifies < WT_MAX_CONSECUTIVE_REVERSE_MODIFY) &&
               __wt_calc_modify(session, prev_full_value, full_value, prev_full_value->size / 10,
                 entries, &nentries) == 0) {
                 WT_ERR(__wt_modify_pack(hs_cursor, entries, nentries, &modify_value));
                 WT_ERR(__hs_insert_record(
                   session, hs_cursor, btree, key, WT_UPDATE_MODIFY, modify_value, &tw));
                 __wt_scr_free(session, &modify_value);
-            } else
+                if (upd->type == WT_UPDATE_STANDARD)
```

Comment:
So the logic is that if the original update is a modify, we will insert it as a reverse modify ignoring the WT_MAX_CONSECUTIVE_REVERSE_MODIFY limit?

Reply:
I think the counter should increase no matter the type is standard or not.

Reply:
That was the behaviour before: if the update was a modify, it could be written as a revere modify.

Now that we are allowing standard update to be written as reverse modify, we want to limit them.

Reply:
So if we have m -> s * 10 -> m, we still can exceed the WT_MAX_CONSECUTIVE_REVERSE_MODIFY limit? I think we should consider modify within the limit as well, otherwise the read performance can suffer in the worst case.

Reply:
> So the logic is that if the original update is a modify, we will insert it as a reverse modify ignoring the WT_MAX_CONSECUTIVE_REVERSE_MODIFY limit?

That was the behaviour before: it could only be written as a reverse modify if the update was a modify.

Reply:
> I think the counter should increase no matter the type is standard or not.

Now that we are allowing standard update to be written as reverse modify, we want to limit the additional reverse modifies that our code changes introduce.

Reply:
> So if we have m -> s * 10 -> m, we still can exceed the WT_MAX_CONSECUTIVE_REVERSE_MODIFY limit? I think we should consider modify within the limit as well, otherwise the read performance can suffer in the worst case.

Yes, in this case we would exceed WT_MAX_CONSECUTIVE_REVERSE_MODIFY.
Before, if we had m * 12, we would write all of them as reverse modifies. Now, if we increment the counter on modify update, we would write one of them as a full update, which will have a direct impact on the HS store size.

Reply:
@quchenhao That's a good point! We did a bunch of testing with both including the modifies in the limit and only enforcing the limit on standard updates. 

We ultimately chose to only enforce the limit on standard updates since we didn't want to change the current behavior of limiting the number of consecutive cursor modifies (since it seemed this was ok to tolerate this read performance-wise?). Additionally for cursor modify heavy workloads it had the adverse effect of _slightly_ increasing the HS size (since 1 in 10 modifies were becoming full updates). This was mainly noticed when testing updates through MongoDB. 
Though your point on having sequence of 10 standard updates between a series of modifies is also valid in impacting read performance. More than happy to include modifies in the count if we feel this is more important!

Reply:
In the previous code, not limiting the number of reverse modifies is OK because we have already done so on the update list so that we can't have more reverse modifies than allowed on the update chain. With this change, in the example I posted, the number of consecutive modifies can be unbounded.

Reply:
Oh, completely missed that detail. 
Yep! it makes sense then to apply the limit to all update types then. We will remove the filter check on standard updates